### PR TITLE
feat(default-memory-limits): set default memory limits 

### PR DESCRIPTION
**Pull request by ryanhristovski**
Fixes https://github.com/envoyproxy/gateway/issues/4882

- sets default memory limits for shutdown manager to 32Mi and default containers to 512Mi ([== to limit](https://github.com/envoyproxy/gateway/issues/4882#issuecomment-2529629806))
- applied [calculated max heap size](https://github.com/envoyproxy/gateway/blob/2a10d474fdfe2f3a9689266355fa6d89a042daac/internal/infrastructure/kubernetes/proxy/resource.go#L356-L369) envoy config to test data for new limit (512Mi)

### CODE REVIEW

File: internal/infrastructure/kubernetes/proxy/testdata/daemonsets/override-labels-and-annotations.yaml [hunk 1]
```diff
@@ -215,6 +215,21 @@ spec:
               typed_config:
                 "@type": type.googleapis.com/envoy.extensions.resource_monitors.downstream_connections.v3.DownstreamConnectionsConfig
                 max_active_downstream_connections: 50000
+            - name: "envoy.resource_monitors.fixed_heap"
```
Comment (zirain) [inline internal/infrastructure/kubernetes/proxy/testdata/daemonsets/override-labels-and-annotations.yaml:218]: why these changes happen?

Comment (ryanhristovski) [reply internal/infrastructure/kubernetes/proxy/testdata/daemonsets/override-labels-and-annotations.yaml:218]: It's set based on the memory limit by default, since memory limits weren't defined before they weren't being applied

Code here https://github.com/envoyproxy/gateway/blob/2a10d474fdfe2f3a9689266355fa6d89a042daac/internal/infrastructure/kubernetes/proxy/resource.go#L356-L369

Comment (zirain) [reply internal/infrastructure/kubernetes/proxy/testdata/daemonsets/override-labels-and-annotations.yaml:218]: thanks for clarify

**Review by zirain — APPROVED** [review]

**Review by cnvergence — APPROVED** [review]
LGTM

**Comment by zirain** [comment]
looks like 32 cause OOM on master.

```
shutdown-manager:
    Container ID:    containerd://e43b953745ffdc5c2e86cf97c061d9e3560b7d800b15b625ca27d4ba315f9ece
    Image:           docker.io/envoyproxy/gateway-dev:latest
    Image ID:        docker.io/envoyproxy/gateway-dev@sha256:cb9737c83359588d3c10683a90ad835ae62696fb3ef8df20383425048aba1133
    Port:            <none>
    Host Port:       <none>
    SeccompProfile:  RuntimeDefault
    Command:
      envoy-gateway
    Args:
      envoy
      shutdown-manager
      --ready-timeout=15s
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       OOMKilled
      Exit Code:    137
      Started:      Sat, 28 Dec 2024 14:15:42 +0800
      Finished:     Sat, 28 Dec 2024 14:15:42 +0800
    Ready:          False
    Restart Count:  5
    Limits:
      memory:  32Mi
    Requests:
      cpu:      10m
      memory:   32Mi
    Liveness:   http-get http://:19002/healthz delay=0s timeout=1s period=10s #success=1 #failure=3
    Readiness:  http-get http://:19002/healthz delay=0s timeout=1s period=10s #success=1 #failure=3
    Startup:    http-get http://:19002/healthz delay=0s timeout=1s period=10s #success=1 #failure=30
    Environment:
      ENVOY_GATEWAY_NAMESPACE:  envoy-gateway-system (v1:metadata.namespace)
      ENVOY_POD_NAME:           envoy-gateway-conformance-infra-all-namespaces-302def45-6brqfc7 (v1:metadata.name)
    Mounts:                     <none>
```

**Comment by ryanhristovski** [comment]
@zirain hmm, it's generally not best practice to have memory request & limits not equal to each other - but given that this is a shutdown manager it _could_ make sense to have separate values.

What do you think:
1. set memory request to 32Mi and memory limit to something like 128Mi
or
2. set both request and limit to 128Mi

I didnt have time to fully benchmark it's resource use but it seems to sit idle around 25Mi and during shutdown (in my environment) it can hit around 70Mi.

I'm leaning #1 due to the nature of the shutdown manager - but this could cause resource issues for users in the future.

**Comment by zirain** [comment]
can you take a look at #4978?

**Comment by zirain** [comment]
@ryanhristovski this block CI, I didn't figure out why(test passed on my local).

let's revert it first, please resubmit later if you still want this, sorry for this.

**Comment by sanposhiho** [comment]
my 2p for the future:
1. Memory consumption depends on Go's GC cycle. In short, if CPU is busy for other stuff, Go runtime cannot take time for GC and hence memory consumption could go up. There's some env parameters (GOMEMLIMIT, GOGC) that can adjust the behaviour though, AFAIK, there's nothing to define a _hard_ limit of memory.
2. CPU `10m` looks too small, and also it doesn't have any limit. No CPU limit isn't a bad thing (rather I'm no-cpu-limit lover), but you have to put CPU limit when benchmarking to measure memory consumption for this kind of change. The reason is (1); obviously, there'd be a difference in memory consumption when your shutdown manager can use 10 CPU vs when your shutdown manager can use only 10m CPU.
